keps/sig-api-machinery/20190802-dynamic-coordinated-storage-version.md

```diff
@@ -0,0 +1,309 @@
+---
+title: StorageVersion API for HA API servers 
+authors:
+  - "@xuchao"
+owning-sig: sig-api-machinery
+reviewers:
+  - "@deads2k"
+  - "@yliaog"
+  - "@lavalamp"
+approvers:
+  - "@deads2k"
+  - "@lavalamp"
+creation-date: 2019-08-22
+last-updated: 2019-08-22
+status: provisional
+---
+
+# StorageVersion API for HA API servers
+
+## Table of Contents
+
+* [Overview](#overview)
+* [API changes](#api-changes)
+   * [Resource Version API](#resource-version-api)
+* [Changes to API servers](#changes-to-api-servers)
+   * [Curating a list of participating API servers in HA master](#curating-a-list-of-participating-api-servers-in-ha-master)
+   * [Updating StorageVersion](#updating-storageversion)
+   * [Garbage collection](#garbage-collection)
+   * [CRDs](#crds)
+   * [Aggregated API servers](#aggregated-api-servers)
+* [Consuming the StorageVersion API](#consuming-the-storageversion-api)
+* [StorageVersion API vs. StorageVersionHash in the discovery document](#storageversion-api-vs-storageversionhash-in-the-discovery-document)
+* [Backwards Compatibility](#backwards-compatibility)
+* [Graduation Plan](#graduation-plan)
+* [FAQ](#faq)
+* [Alternatives](#alternatives)
+   * [Letting API servers vote on the storage version](#letting-api-servers-vote-on-the-storage-version)
+   * [Letting the storage migrator detect if API server instances are in agreement](#letting-the-storage-migrator-detect-if-api-server-instances-are-in-agreement)
+* [Appendix](#appendix)
+   * [Accuracy of the discovery document of CRDs](#accuracy-of-the-discovery-document-of-crds)
+* [References](#references)
+
+## Overview
+
+During the rolling upgrade of an HA master, the API server instances may
+use different storage versions encoding a resource. The [storageVersionHash][]
+in the discovery document does not expose this disagreement. As a result, the
+storage migrator may proceed with migration with the false belief that all API
+server instances are encoding objects using the same storage version, resulting
+in polluted migration.  ([details][]).
+
+[storageVersionHash]:https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L979
+[details]:https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/35-storage-version-hash.md#ha-masters
+
+We propose a mechanism for HA API servers to expose if they have unanimously
+agreed on the storage version, so that the storage migrator can defer migration
+until an agreement has been reached.
+
+## API changes
+
+### Resource Version API
+
+We introduce a new API `StorageVersion`, in a new API group
+`apiserver.internal.k8s.io/v1alpha1`. 
```

Comment:
> I think it's OK to use internal, but I think it should be `internal.apiserver.k8s.io`. Also is k8s.io the best place to put it? Is there something less conspicuous, like sig-scoped rather than org-scoped?

I think it would be a little weird for the kube-apiserver to be natively setting an `x-k8s.io` resource.  I don't think the core-k8s should be directly relying on non-apireviewed APIs.